make/compiler_flags

```diff
@@ -0,0 +1,235 @@
+# unassigned operator (`?=`) when it's not set.
+##
+O ?= 3
+
+## setup includes
+INC ?= $(INC_FIRST) -I $(if $(MATH),$(MATH),.) -I $(EIGEN) -I $(BOOST) $(INC_SUNDIALS)
+INC_SUNDIALS ?= -I $(SUNDIALS)/include
+INC_GTEST ?= -I $(GTEST)/include -I $(GTEST)
+
+## setup precompiler options
+CPPFLAGS_BOOST ?= -DBOOST_RESULT_OF_USE_TR1 -DBOOST_NO_DECLTYPE -DBOOST_DISABLE_ASSERTS -DBOOST_PHOENIX_NO_VARIADIC_EXPRESSION
+CPPFLAGS_SUNDIALS ?= -DNO_FPRINTF_OUTPUT
+CPPFLAGS_GTEST ?= -DGTEST_USE_OWN_TR1_TUPLE
+
+
+## setup compiler flags
+CXXFLAGS_LANG ?= -std=c++1y
+#CXXFLAGS_BOOST
+CXXFLAGS_SUNDIALS ?= -pipe
+#CXXFLAGS_GTEST
+
+
+## These are the other compiler flags that can be set.
+##
+## They are listed here for documentation and are intentionally
+## commented out.
+
+# CXXFLAGS_OS
+# CXXFLAGS_WARNINGS
+# CXXFLAGS_EIGEN
+
+## GPU specific flags
+# CXXFLAGS_OPENCL
+# LDFLAGS_OPENCL
+# LDLIBS_OPENCL
+
+## MPI specific flags
+# CXXFLAGS_MPI
+# LDFLAGS_MPI
+# LDLIBS_MPI
+
+################################################################################
+# Update compiler flags with operating system specific modifications
+##
+ifeq ($(OS),Windows_NT)
+  CXXFLAGS_WARNINGS ?= -Wall -Wno-unused-function -Wno-uninitialized -Wno-unused-but-set-variable -Wno-unused-variable -Wno-sign-compare -Wno-unused-local-typedefs
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CPPFLAGS_OS ?= -D_USE_MATH_DEFINES
+  ## Detecting Process Bitness:
+  ## http://blogs.msdn.com/b/david.wang/archive/2006/03/26/howto-detect-process-bitness.aspx
+  ifeq ($(PROCESSOR_ARCHITECTURE)$(PROCESSOR_ARCHITEW6432),x86)
+    CXXFLAGS_OS ?= -m32
+  else
+    CXXFLAGS_OS ?= -m64
+  endif
+
+  ifneq (gcc,$(CXX_TYPE))
+    LDLIBS_OS ?= -static-libgcc
+  else
+    LDLIBS_OS ?= -static-libgcc -static-libstdc++
+  endif
+
+  EXE := .exe
+endif
+
+ifeq ($(OS),Darwin)
+  ifeq (clang,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -Wno-unknown-warning-option -Wno-tautological-compare -Wno-sign-compare
+  endif
+endif
+
+ifeq ($(OS),Linux)
+  CPPFLAGS_GTEST ?= -DGTEST_HAS_PTHREAD=0
+  CXXFLAGS_WARNINGS ?= -Wno-sign-compare
+  ifeq (gcc,$(CXX_TYPE))
+    CXXFLAGS_OS ?= -pthread
+  endif
+  ifeq (mingw32-gcc,$(CXX_TYPE))
+    LDLIBS_OS ?= -static -lwinpthread
+    EXE := .exe
+    WINE ?= wine
+  endif
+  ifeq (clang,$(CXX_TYPE))
+    LDLIBS_OS ?= -lpthread
+  endif
+endif
+
+
+################################################################################
+# Setup GPU
+#
+# Adds the following to CXXFLAGS_OPENCL
+# OPENCL_DEVICE_ID - The ID of the GPU (default: 0)
+# OPENCL_PLATFORM_ID The ID of the OpenCL platform (default: 0)
+# Both IDs can be found through installing and calling clinfo
+
+ifdef STAN_OPENCL
+  ifeq ($(OS),Darwin)  # Mac OS X
+    LDLIBS_OPENCL ?= -framework OpenCL
+  else
+    LDLIBS_OPENCL ?= -lOpenCL
+  endif
+  OPENCL_DEVICE_ID ?= 0
+  OPENCL_PLATFORM_ID ?= 0
+
+  CPPFLAGS_OPENCL ?= -DSTAN_OPENCL -DOPENCL_DEVICE_ID=$(OPENCL_DEVICE_ID) -DOPENCL_PLATFORM_ID=$(OPENCL_PLATFORM_ID)
```

Comment:
I am now seeing compiler warnings on deprecated OpenCL calls, that were not there before. Would if be possible to add `-DCL_USE_DEPRECATED_OPENCL_1_2_APIS` here?

Its the same issue as [here](https://stackoverflow.com/questions/28500496/opencl-function-found-deprecated-by-visual-studio) but its the same on Linux and Windows.

We are using OpenCL 1.2, we will switch to 2.0 eventually, but in order to ensure that we support as much hardware as possible.

Reply:
The warnings look like this: 

```
In file included from ./stan/math/prim/arr/err/check_opencl.hpp:7:0,
                 from ./stan/math/prim/arr.hpp:18,
                 from ./stan/math/prim/mat.hpp:325,
                 from ./stan/math/rev/mat.hpp:12,
                 from ./stan/math.hpp:4,
                 from test/unit/math_include_test.cpp:1:
lib/opencl_1.2.8/CL/cl.hpp: In constructor 'cl::Sampler::Sampler(const cl::Context&, cl_bool, cl_addressing_mode, cl_filter_mode, cl_int*)':
lib/opencl_1.2.8/CL/cl.hpp:4645:21: warning: '_cl_sampler* clCreateSampler(cl_context, cl_bool, cl_addressing_mode, cl_filter_mode, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1450) [-Wdeprecated-declarations]
         object_ = ::clCreateSampler(
                     ^
lib/opencl_1.2.8/CL/cl.hpp:4650:19: warning: '_cl_sampler* clCreateSampler(cl_context, cl_bool, cl_addressing_mode, cl_filter_mode, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1450) [-Wdeprecated-declarations]
             &error);
                   ^
In file included from ./stan/math/prim/arr/err/check_opencl.hpp:7:0,
                 from ./stan/math/prim/arr.hpp:18,
                 from ./stan/math/prim/mat.hpp:325,
                 from ./stan/math/rev/mat.hpp:12,
                 from ./stan/math.hpp:4,
                 from test/unit/math_include_test.cpp:1:
lib/opencl_1.2.8/CL/cl.hpp: In constructor 'cl::CommandQueue::CommandQueue(cl_command_queue_properties, cl_int*)':
lib/opencl_1.2.8/CL/cl.hpp:5505:25: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
             object_ = ::clCreateCommandQueue(
                         ^
lib/opencl_1.2.8/CL/cl.hpp:5506:56: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
                 context(), device(), properties, &error);
                                                        ^
lib/opencl_1.2.8/CL/cl.hpp: In constructor 'cl::CommandQueue::CommandQueue(const cl::Context&, cl_command_queue_properties, cl_int*)':
lib/opencl_1.2.8/CL/cl.hpp:5536:21: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
         object_ = ::clCreateCommandQueue(context(), devices[0](), properties, &error);
                     ^
lib/opencl_1.2.8/CL/cl.hpp:5536:85: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
         object_ = ::clCreateCommandQueue(context(), devices[0](), properties, &error);
                                                                                     ^
lib/opencl_1.2.8/CL/cl.hpp: In constructor 'cl::CommandQueue::CommandQueue(const cl::Context&, const cl::Device&, cl_command_queue_properties, cl_int*)':
lib/opencl_1.2.8/CL/cl.hpp:5553:21: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
         object_ = ::clCreateCommandQueue(
                     ^
lib/opencl_1.2.8/CL/cl.hpp:5554:52: warning: '_cl_command_queue* clCreateCommandQueue(cl_context, cl_device_id, cl_command_queue_properties, cl_int*)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1443) [-Wdeprecated-declarations]
             context(), device(), properties, &error);
                                                    ^
lib/opencl_1.2.8/CL/cl.hpp: In member function 'cl_int cl::CommandQueue::enqueueTask(const cl::Kernel&, const std::vector<cl::Event>*, cl::Event*) const':
lib/opencl_1.2.8/CL/cl.hpp:6357:15: warning: 'cl_int clEnqueueTask(cl_command_queue, cl_kernel, cl_uint, _cl_event* const*, _cl_event**)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1457) [-Wdeprecated-declarations]
             ::clEnqueueTask(
               ^
lib/opencl_1.2.8/CL/cl.hpp:6361:46: warning: 'cl_int clEnqueueTask(cl_command_queue, cl_kernel, cl_uint, _cl_event* const*, _cl_event**)' is deprecated (declared at lib/opencl_1.2.8/CL/cl.h:1457) [-Wdeprecated-declarations]
                 (event != NULL) ? &tmp : NULL),
```